digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/process/TrackSectionOrder.scala

```diff
@@ -0,0 +1,133 @@
+
+import fi.liikennevirasto.digiroad2.{GeometryUtils, Matrix, Point, Vector3d}
+import fi.liikennevirasto.digiroad2.asset.SideCode
+import fi.liikennevirasto.digiroad2.asset.SideCode.{AgainstDigitizing, TowardsDigitizing}
+import fi.liikennevirasto.digiroad2.util.{RoadAddressException, Track}
+import fi.liikennevirasto.digiroad2.util.Track.RightSide
+import fi.liikennevirasto.viite.dao.ProjectLink
+
+
+object TrackSectionOrder {
+  private val RightVector = Vector3d(-1.0, 0.0, 0.0)
+  private val ForwardVector = Vector3d(0.0, 1.0, 0.0)
+
+  def RotationMatrix(tangent: Vector3d): Matrix = {
+    if (Math.abs(tangent.x) <= fi.liikennevirasto.viite.Epsilon)
+      Matrix(Seq(Seq(0.0, -1.0), Seq(1.0, 0.0)))
+    else {
+      val k = tangent.y / tangent.x
+      val coeff = 1/Math.sqrt(k*k + 1)
+      Matrix(Seq(Seq(coeff, -k*coeff), Seq(k*coeff, coeff)))
+    }
+  }
+
+  def orderProjectLinksTopologyByGeometry(startingPoints: (Point, Point), list: Seq[ProjectLink]): (Seq[ProjectLink], Seq[ProjectLink]) = {
+
+    def findOnceConnectedLinks(seq: Seq[ProjectLink]): Seq[ProjectLink] = {
+      // Using 3 because case b==j is included
+      seq.filter(b => seq.count(j =>
+        GeometryUtils.areAdjacent(b.geometry, j.geometry, fi.liikennevirasto.viite.MaxDistanceForConnectedLinks)) < 3)
+    }
+    def distancePointToLink(p: Point, l: ProjectLink) = {
+      GeometryUtils.minimumDistance(p, GeometryUtils.geometryEndpoints(l.geometry))
+    }
+    def pickMostAligned(rotationMatrix: Matrix, vector: Vector3d, candidates: Seq[ProjectLink]): ProjectLink = {
+      candidates.minBy(pl => (rotationMatrix* GeometryUtils.firstSegmentDirection(pl.geometry).normalize2D()).dot(vector))
+    }
+
+    def pickRightMost(lastLink: ProjectLink, candidates: Seq[ProjectLink]): ProjectLink = {
+      pickMostAligned(RotationMatrix(GeometryUtils.lastSegmentDirection(lastLink.geometry)), RightVector, candidates)
+    }
+
+    def pickForwardPointing(lastLink: ProjectLink, candidates: Seq[ProjectLink]): ProjectLink = {
+      pickMostAligned(RotationMatrix(GeometryUtils.lastSegmentDirection(lastLink.geometry)), ForwardVector, candidates)
+    }
+
+    def recursiveFindAndExtend(currentPoint: Point, ready: Seq[ProjectLink], unprocessed: Seq[ProjectLink]): Seq[ProjectLink] = {
+      if (unprocessed.isEmpty)
+        ready
+      else {
+        val connected = unprocessed.filter(pl => GeometryUtils.minimumDistance(currentPoint,
+          GeometryUtils.geometryEndpoints(pl.geometry)) < fi.liikennevirasto.viite.MaxDistanceForConnectedLinks)
+        val nextLink = connected.size match {
+          case 0 =>
+            val subsetB = findOnceConnectedLinks(unprocessed)
+            subsetB.minBy(b => distancePointToLink(currentPoint, b))
+          case 1 =>
+            connected.head
+          case 2 =>
+            if (findOnceConnectedLinks(unprocessed).exists(b =>
+              distancePointToLink(currentPoint, b) <= fi.liikennevirasto.viite.MaxJumpForSection)) {
```

Comment:
should "<="  be "<"?

Reply:
ah, nvm ≤ 50 in algorithm

Reply:
"etsitään joukosta J alijoukko B, jonka linkit ovat korkeintaan 50 metrin päässä ja jonka linkeillä on korkeintaan yksi liittyvä linkki joukossa J: b ∈ B ↔ |b - p| ≤ 50 ∧ ∃! j ∈ J: min |b - j| ≤ 0.1"

-> no.